storage/memory.go

```diff
@@ -0,0 +1,52 @@
+package storage
+
+import (
+	"bytes"
+	"fmt"
+	"io"
+
+	"github.com/ipld/go-ipld-prime"
+)
+
+// Memory is a simple in-memory storage for data indexed by ipld.Link.
+// (It's little more than a map -- in fact, the map is exported,
+// and you can poke it directly.)
+//
+// The OpenRead method conforms to ipld.BlockReadOpener,
+// and the OpenWrite method conforms to ipld.BlockWriteOpener.
+// Therefore it's easy to use in a LinkSystem like this:
+//
+//		store := storage.Memory{}
+//		lsys.StorageReadOpener = (&store).OpenRead
+//		lsys.StorageWriteOpener = (&store).OpenWrite
```

Comment:
I find it a bit weird to have to supply a storage reader and a storage writer as separate entities. When supplying both, it's going to be very common that they'll both come from the same implementation, like storage.Memory here.

So it would seem more natural to me to define an interface, and be able to supply both read and write ends in one go. For example:

```
store := storage.Memory{}
lsys = lsys.WithStore(store)
```

Reply:
I'd like to try this way.  It's close to what we've got already; and are several reasons I'd like to use functional interfaces and be able to set these separately:

Reason 1: It's true that when supplying both, they often come from the same implementation... but I really like *not having to supply both*.  **Making something read-only by not providing it an ability to write sits well with me.**  It's a "capabilities" style system, and I like those: the lines are very clear.

Reason 2: Functions as interfaces makes it easier to decorate one or the other (or both), because when they're just functions, it can be done inline.  By contrast, a whole type that has methods attached to satisfy an interface has to be a top-level declaration.  It's also harder to decorate just one of the several functions on an interface; doing so tends to cause one to end up with a whole suite of decorator wrapper types.  I think it's been not uncommon so far for us to want to decorate one of these methods but not care much about the other.

Reason 3: An interface, because of how it has the ability to grow, makes it more easier for something to evolve to become stateful.  I think that would be a bad direction of growth for these data in and out primitives.  These should be really simple and really functional.

These reasons 2 and 3 may sound a little hypothetical, but I think it's what we can observe actually happening in the previous generation of interfaces for this domain: [go-ipfs-blockstore](https://github.com/ipfs/go-ipfs-blockstore/blob/7adc396ab4b40c660c72d215b3a58999b6f63413/blockstore.go#L35) grew in both these directions over time, and while the result hasn't been _wrong_ per se, it also hasn't been very satisfying and it hasn't stayed simple; I seem to hear grumbling from people who look at that interface for long.  I don't want to set ourselves up for drift in that same direction again.

Reason 4: This pattern just plain has worked well for me in the past, and I think functional interfaces are underutilized in golang :D

I'd revisit this if we did a design review in depth on the blockstore concepts, and found with confidence that we wanted a certain sequence of methods to always be available in a bundle together; or, if we have some ideas where we'd actually want to be able to do feature detection on the storage system.

At the moment, my thinking is that I don't think we'll *want* the ability to do further feature detection on these values in any of the places we use a LinkSystem.  For example: will the traversal package ever want to check if the storage system supports GC (which is one of the other interfaces that have appeared around blockstorage)?  I think it shouldn't.

The best example I can think of for wanting feature detection is probably the https://pkg.go.dev/github.com/ipfs/go-ipfs-blockstore#Viewer interface.  But, we've also seen that to be doable by checking the `io.Reader` *returned* by these functions for other features.  So we probably don't need to be preparing to do things like that in two separate places.

Reply:
Maybe a middle position on this -- let's try functional interfaces for now (this is also the smallest change from the current `Loader` and `Storer` system); but we stay extremely ready to change this as folks work on moving go-ipfs over to these libraries?  We should get good and hard info from that process, and if we discover the need for more feature detection on storage systems during that work, then we do it.  We could defer tagging a new version until we've got enough experience from that work to decide more confidently.